opnode/p2p/host_test.go

```diff
@@ -0,0 +1,136 @@
+	yamux "github.com/libp2p/go-libp2p-yamux"
+	lconf "github.com/libp2p/go-libp2p/config"
+	mocknet "github.com/libp2p/go-libp2p/p2p/net/mock"
+	"github.com/stretchr/testify/require"
+)
+
+func TestingConfig(t *testing.T) *Config {
+	p, _, err := crypto.GenerateSecp256k1Key(rand.Reader)
+	require.NoError(t, err, "failed to generate new p2p priv key")
+	mtpt, err := lconf.MuxerConstructor(yamux.DefaultTransport)
+	require.NoError(t, err)
+	mux := lconf.MsMuxC{MuxC: mtpt, ID: "/yamux/1.0.0"}
+
+	return &Config{
+		Priv:                (*ecdsa.PrivateKey)((p).(*crypto.Secp256k1PrivateKey)),
+		DisableP2P:          false,
+		NoDiscovery:         true, // we statically peer during most tests.
+		ListenIP:            net.IP{127, 0, 0, 1},
+		ListenTCPPort:       0, // bind to any available port
+		StaticPeers:         nil,
+		HostMux:             []lconf.MsMuxC{mux},
+		NoTransportSecurity: true,
+		PeersLo:             1,
+		PeersHi:             10,
+		PeersGrace:          time.Second * 10,
+		NAT:                 false,
+		UserAgent:           "optimism-testing",
+		TimeoutNegotiation:  time.Second * 2,
+		TimeoutAccept:       time.Second * 2,
+		TimeoutDial:         time.Second * 2,
+		Store:               sync.MutexWrap(ds.NewMapDatastore()),
+		ConnGater: func(conf *Config) (connmgr.ConnectionGater, error) {
+			return tswarm.DefaultMockConnectionGater(), nil
+		},
+		ConnMngr: DefaultConnManager,
+	}
+}
+
+// Simplified p2p test, to debug/test basic libp2p things with
+func TestP2PSimple(t *testing.T) {
+	confA := TestingConfig(t)
+	confB := TestingConfig(t)
+	hostA, err := confA.Host()
+	require.NoError(t, err, "failed to launch host A")
+	defer hostA.Close()
+	hostB, err := confB.Host()
+	require.NoError(t, err, "failed to launch host B")
+	defer hostB.Close()
+	err = hostA.Connect(context.Background(), peer.AddrInfo{ID: hostB.ID(), Addrs: hostB.Addrs()})
+	require.NoError(t, err, "failed to connect to peer B from peer A")
+	require.Equal(t, hostB.Network().Connectedness(hostA.ID()), network.Connected)
+}
+
+// Full setup, using negotiated transport security and muxes
+func TestP2PFull(t *testing.T) {
+	pA, _, err := crypto.GenerateSecp256k1Key(rand.Reader)
+	require.NoError(t, err, "failed to generate new p2p priv key")
+	pB, _, err := crypto.GenerateSecp256k1Key(rand.Reader)
+	require.NoError(t, err, "failed to generate new p2p priv key")
+
+	mplexC, err := mplexC()
+	require.NoError(t, err)
+	yamuxC, err := yamuxC()
+	require.NoError(t, err)
+	noiseC, err := noiseC()
+	require.NoError(t, err)
+	tlsC, err := tlsC()
+	require.NoError(t, err)
+
+	confA := Config{
+		Priv:                (*ecdsa.PrivateKey)((pA).(*crypto.Secp256k1PrivateKey)),
+		DisableP2P:          false,
+		NoDiscovery:         true,
+		ListenIP:            net.IP{127, 0, 0, 1},
+		ListenTCPPort:       0, // bind to any available port
+		StaticPeers:         nil,
+		HostMux:             []lconf.MsMuxC{yamuxC, mplexC},
+		HostSecurity:        []lconf.MsSecC{noiseC, tlsC},
```

Comment:
For both `HostSecurity` and `HostMux`, what do you think of limiting each of these to only one option? Concerned about offering too many different configurations that have to be negotiated at connection time. We have the ability to require specific settings since the network isn't live yet.

Reply:
I prefer to have at least the two most popular options. When one of them breaks or underperforms we can prioritize the other in an upgrade, and keep the old and new nodes connected.